### DISCUSSION TITLE
Making a modification to Device found in Stylo

### DISCUSSION THREAD

**Discussion by arthmis** [servo/servo#34516] [Q&A] [answered]
I'm working on [this issue](https://github.com/servo/servo/issues/34425) and I've reached a roadblock. I want to modify the `Device` that the `LayoutThread` owns, through Stylist, in order to change the `PrefersColorScheme` [found here](https://github.com/servo/stylo/blob/35e0fa29e93a7e85f44d4ad0610748634f7979b5/style/servo/media_queries.rs#L92). Right now I'm thinking `Device` needs some way to mutate it from the `LayoutThread` but I'm not very sure of that. Does that make sense? I can try and provide more context.

@jdm, @gterzian, @sagudev, or @mrobinson

**Comment by mrobinson** [comment]
@nicoburns is working on this issue, so he is probably the best one to answer this question.

**Comment by arthmis** [comment]
Oh yeah I did ask Nico about it so it would be good to get him in the discussion, thanks. Nico did mention that there is a [`handle_viewport_change`](https://github.com/servo/servo/blob/af8f35f5ef51e9743274f149900f0c9c80ca5288/components/layout_thread_2020/lib.rs#L1110) function that might provide insight on how I can tackle this issue. However, in that example there isn't a way for the `Device` to retain it's current color scheme.

I'm thinking that there should be a function like `fn theme(&self) -> PrefersColorScheme` for `Device` to provide current color scheme when recreating a new device for the `handle_viewport_change` function.

As for responding to the color scheme change event from the user's device/OS I am thinking there should be some functionality to modify a device or recreate a device directly.

 @mrobinson  Then after setting the color scheme, would a reflow be required? This part I'm also unsure of because I don't know if there is an expectation of the layout changing when user color scheme preferences changes.

**Reply by nicoburns** [reply, marked as answer]
> I'm thinking that there should be a function like fn theme(&self) -> PrefersColorScheme for Device to provide current color scheme when recreating a new device for the handle_viewport_change function.

This is what I had in mind in the short term (and you might need similar for the viewport size). We are also looking into making it possible to mutate the Device directly. But I think that ought to be a separate change.

**Reply by arthmis** [reply]
oh ok, if you're fine with this in the short term then I think have a working solution. I'm going to open a draft PR. There is a particular area concerning `pipeline_ids` that I'm not sure about.

**Comment by arthmis** [comment]
I actually figured out the idea Nico was suggesting by making the theme change in the `Device` through a reflow within `handle_viewport_change`. I'm currently trying to test the feature and will probably make a draft PR.

**Reply by arthmis** [reply]
Draft PR has been made https://github.com/servo/servo/pull/34532 @nicoburns
